### PR TITLE
Entity cache: configuration

### LINKED ISSUE #4194: Entity caching: configuration
support global and per subgraph configuration. In particular, TTL should be modifiable per subgraph

### PR DESCRIPTION
**Pull request by Geal**
Fix #4194

```yaml title="router.yaml"
experimental_entity_cache:
  redis:
    urls: [ "redis://localhost:6379" ]
    timeout: 5ms # Optional, by default: 2ms
    ttl: 60s # Optional, by default no expiration
  enabled: true
  subgraphs:
    accounts:
      enabled: false
    products:
      ttl: 120s
```
**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [ ] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [ ] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [ ] Unit Tests
    - [ ] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

### CODE REVIEW

**Review by bnjjj — APPROVED** [review]
Maybe we should now rename `timeout` to `connection_timeout` to avoid confusion with `ttl` ?

**Review by BrynCooke — COMMENTED** [review]
Looks like config metrics are missing. 
Will this be in a follow up PR?

**Comment by Geal** [comment]
>Looks like config metrics are missing.
Will this be in a follow up PR?

good point, it comes in https://github.com/apollographql/router/pull/4226/commits/c9c865b39dceb2899b1684bb2e6863dba01e0e8d. Would it be worth it to expose the usage of the ttl field?

**Comment by Geal** [comment]
adding the `ttl` and `enabled` options to metrics in ccce76daa

**Comment by BrynCooke** [comment]
I think we should have some way of specifying a key prefix. Users may be using a redis cluster for several purposes, and being able to have a custom prefix would be beneficial? I could be wrong, and if there is another way of doing this then disregard this.

**Comment by Geal** [comment]
you're right, a key prefix would make sense (for APQ and query planning caches too), I even opend an issue about it yesterday https://github.com/apollographql/router/issues/4247 :)
I'll implement that shortly

File: apollo-router/src/plugins/cache/entity.rs [hunk 1]
```diff
@@ -40,15 +41,44 @@ register_plugin!("apollo", "experimental_entity_cache", EntityCache);
...
+    enabled: Option<bool>,
+    /// Per subgraph configuration
+    #[serde(default)]
+    subgraphs: HashMap<String, Subgraph>,
 }
 
+/// Per subgraph configuration for entity caching
+#[derive(Clone, Debug, JsonSchema, Deserialize)]
+#[serde(rename_all = "snake_case", deny_unknown_fields)]
+struct Subgraph {
```
Comment (BrynCooke) [inline apollo-router/src/plugins/cache/entity.rs:64]: Nit could we get away with serde default at the struct level and implement `Default` for subgraph?

A long TTL and enabled true?

**Review by BrynCooke — APPROVED** [review]

File: apollo-router/src/plugins/cache/entity.rs (hunk 1 above)
Comment (Geal) [reply apollo-router/src/plugins/cache/entity.rs:64]: This is one of those cases where the subgraph config has to override the global config, but only for the fields specified in the subgraph, so I can't set default values

**Pull request merged**